internal/controller/dragonfly_controller.go

```diff
@@ -88,7 +89,127 @@ func (r *DragonflyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			return ctrl.Result{}, err
 		}
 
-		r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Created", "Created resources for Dragonfly object")
+		r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Resources", "Created resources")
+		return ctrl.Result{}, nil
+	} else if df.Status.Phase == PhaseRollout {
+		// This is a Rollout
+		log.Info("Rolling out new version")
+		var updatedStatefulset appsv1.StatefulSet
+		if err := r.Get(ctx, client.ObjectKey{Namespace: df.Namespace, Name: df.Name}, &updatedStatefulset); err != nil {
+			log.Error(err, "could not get statefulset")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		// get pods of the statefulset
+		var pods corev1.PodList
+		if err := r.List(ctx, &pods, client.InNamespace(df.Namespace), client.MatchingLabels(map[string]string{
+			"app":                               df.Name,
+			resources.KubernetesAppNameLabelKey: "dragonfly",
+		})); err != nil {
+			log.Error(err, "could not list pods")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		if len(pods.Items) != int(*updatedStatefulset.Spec.Replicas) {
+			log.Info("Waiting for all replicas to be ready")
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// filter replicas
+		var master corev1.Pod
+		replicas := make([]corev1.Pod, 0)
+		for _, pod := range pods.Items {
+			if _, ok := pod.Labels[resources.Role]; ok {
+				if pod.Labels[resources.Role] == resources.Replica {
+					replicas = append(replicas, pod)
+				} else if pod.Labels[resources.Role] == resources.Master {
+					master = pod
+				}
+			} else {
+				log.Info("found pod without label", "pod", pod.Name)
+				// retry after they are ready
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+			}
+		}
+
+		// Check if all the latest replicas had a full sync
+		// if not, requeue
+		fullSyncedUpdatedReplicas := 0
+		for _, replica := range replicas {
+			// Check only with latest replicas
+			onLatestVersion, err := isPodOnLatestVersion(ctx, r.Client, &replica, &updatedStatefulset)
+			if err != nil {
+				log.Error(err, "could not check if pod is on latest version")
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, err
+			}
+			if onLatestVersion {
+				// check if the replica had a full sync
+				log.Info("New Replica found. Checking if replica had a full sync", "pod", replica.Name)
+				err := checkForStableSync(ctx, r.Client, &replica)
```

Comment:
I find checkForStableSync rerunning error only a bit confusing, it is not an error state to have a full sync in progress.
I'd go with something like
full_sync_in_progress = !checkForStableSync
...

if full_sync_in_progress {
  log.Info("Full sync in progress for an updated replica, requeuing...")
  return ctrl.Result{RequeueAfter: 5 * time.Second}
}

Reply:
We initially had a boolean state like you suggested, but with that as we check for multiple conditions inside `checkForStableState`, The condition here didn't really know what was wrong and it wasn't propogated in logs (as we don't pass a logger there). Hence, Changed it like this. It kind-off doesn't really matter if its an error state as we are handling it unlike a panic or something. Happy to update it back though if it makes sense.